x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -685,19 +686,54 @@ protected LogicalPlan rule(LogicalPlan e) {
 
         @Override
         protected LogicalPlan rule(Filter filter) {
-            if (filter.condition() instanceof Literal) {
-                if (TRUE.equals(filter.condition())) {
+            Expression condition = filter.condition().transformDown(FoldBinaryLogical.INSTANCE);
+
+            if (condition instanceof Literal) {
+                if (TRUE.equals(condition)) {
                     return filter.child();
                 }
-                if (FALSE.equals(filter.condition()) || Expressions.isNull(filter.condition())) {
+                if (FALSE.equals(condition) || Expressions.isNull(condition)) {
                     return new LocalRelation(filter.location(), new EmptyExecutable(filter.output()));
                 }
             }
 
+            if (!condition.equals(filter.condition())) {
+                return new Filter(filter.location(), filter.child(), condition);
+            }
             return filter;
         }
     }
 
+    static class FoldBinaryLogical implements java.util.function.Function<Expression, Expression> {
+
+        static FoldBinaryLogical INSTANCE = new FoldBinaryLogical();
+
+        @Override
+        public Expression apply(Expression expression) {
```

Comment:
The issue is that `Or` and `And` cannot be folded in the `FoldNull` (except if both are null) because we need the 3vl in the select clause. So for example: we cannot fold `col1 or null` there, because this can potentially return `true` or `null`, for `and`: se cannot fold `col1 and null` because depending on col1 in can result to `false` or `null`. But when we are inside the PruneFilter we know that now we can fold.